File: mip_reference/node-express/lib/mip/models/node-identity.js
'use strict';

const mipCrypto = require('../crypto');
const identifier = require('../identifier');

class NodeIdentity {
  constructor(attrs = {}) {
    this.mipIdentifier = attrs.mipIdentifier;
    this.privateKey = attrs.privateKey;
    this.publicKey = attrs.publicKey;
    this.organizationName = attrs.organizationName;
    this.contactPerson = attrs.contactPerson;
    this.contactPhone = attrs.contactPhone;
    this.mipUrl = attrs.mipUrl;
    this.shareMyOrganization = attrs.shareMyOrganization !== undefined ? attrs.shareMyOrganization : true;
    this.trustThreshold = attrs.trustThreshold !== undefined ? attrs.trustThreshold : 1;
    this.port = attrs.port;
  }

  publicKeyFingerprint() {
    return mipCrypto.fingerprint(this.publicKey);
  }

  toNodeProfile() {
    return {
      mip_identifier: this.mipIdentifier,
      mip_url: this.mipUrl,
      organization_legal_name: this.organizationName,
      contact_person: this.contactPerson,
      contact_phone: this.contactPhone,
      public_key: this.publicKey,
      share_my_organization: this.shareMyOrganization
    };
  }

  // Generate a new node identity from config
  static fromConfig(config, port) {
    const keys = mipCrypto.generateKeyPair();
    const mipId = identifier.generate(config.organization_name);

    return new NodeIdentity({
      mipIdentifier: mipId,
      privateKey: keys.privateKey,
      publicKey: keys.publicKey,
      organizationName: config.organization_name,
      contactPerson: config.contact_person,
      contactPhone: config.contact_phone,
      mipUrl: `http://localhost:${port}/mip/node/${mipId}`,
      shareMyOrganization: config.share_my_organization !== undefined ? config.share_my_organization : true,
      trustThreshold: config.trust_threshold !== undefined ? config.trust_threshold : 1,
      port
    });
  }
}

module.exports = NodeIdentity;

File: mip-ruby/lib/member_interchange_protocol/models/node_identity.rb
# frozen_string_literal: true

module MemberInterchangeProtocol
  module Models
    # Represents this node's identity in the MIP network
    class NodeIdentity
      attr_reader :mip_identifier, :private_key, :public_key, :organization_name,
                  :contact_person, :contact_phone, :mip_url, :share_my_organization,
                  :trust_threshold, :port

      def initialize(attrs = {})
        @mip_identifier = attrs[:mip_identifier]
        @private_key = attrs[:private_key]
        @public_key = attrs[:public_key]
        @organization_name = attrs[:organization_name]
        @contact_person = attrs[:contact_person]
        @contact_phone = attrs[:contact_phone]
        @mip_url = attrs[:mip_url]
        @share_my_organization = attrs.fetch(:share_my_organization, true)
        @trust_threshold = attrs.fetch(:trust_threshold, 1)
        @port = attrs[:port]
      end

      def public_key_fingerprint
        MemberInterchangeProtocol::Crypto.fingerprint(@public_key)
      end

      def to_node_profile
        {
          mip_identifier: @mip_identifier,
          mip_url: @mip_url,
          organization_legal_name: @organization_name,
          contact_person: @contact_person,
          contact_phone: @contact_phone,
          public_key: @public_key,
          share_my_organization: @share_my_organization
        }
      end

      # Generate a new node identity with fresh keys
      def self.generate(organization_name:, contact_person: nil, contact_phone: nil,
                        mip_url: nil, share_my_organization: true, trust_threshold: 1, port: nil)
        keys = MemberInterchangeProtocol::Crypto.generate_key_pair
        mip_id = MemberInterchangeProtocol::Identifier.generate(organization_name)

        new(
          mip_identifier: mip_id,
          private_key: keys[:private_key],
          public_key: keys[:public_key],
          organization_name: organization_name,
          contact_person: contact_person,
          contact_phone: contact_phone,
          mip_url: mip_url || "http://localhost:#{port || 4000}/mip/node/#{mip_id}",
          share_my_organization: share_my_organization,
          trust_threshold: trust_threshold,
          port: port
        )
      end

      # Generate a new node identity from config hash
      def self.from_config(config, port)
        keys = MemberInterchangeProtocol::Crypto.generate_key_pair
        mip_id = MemberInterchangeProtocol::Identifier.generate(config['organization_name'])

        new(
          mip_identifier: mip_id,
          private_key: keys[:private_key],
          public_key: keys[:public_key],
          organization_name: config['organization_name'],
          contact_person: config['contact_person'],
          contact_phone: config['contact_phone'],
          mip_url: "http://localhost:#{port}/mip/node/#{mip_id}",
          share_my_organization: config.fetch('share_my_organization', true),
          trust_threshold: config.fetch('trust_threshold', 1),
          port: port
        )
      end
    end
  end
end
